pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -268,44 +268,42 @@ private void addInstanceGroupTagIfNeeded() {
    */
 
   /**
-   * Get all instance Ids.
-   *
+   * Get all instance Ids. If it's related to segment operation like assigning segments to instances, cache accessor would be used.
+   * Otherwise, direct data accessor would be used.
+   * @param segmentRelatedOperation whether it's related to segment operation
    * @return List of instance Ids
    */
   @Nonnull
-  public List<String> getAllInstances() {
-    return _cacheInstanceConfigsDataAccessor.getChildNames("/", AccessOption.PERSISTENT);
+  public List<String> getAllInstances(boolean segmentRelatedOperation) {
+    if (segmentRelatedOperation) {
+      return _cacheInstanceConfigsDataAccessor.getChildNames("/", AccessOption.PERSISTENT);
+    } else {
+      return _helixAdmin.getInstancesInCluster(_helixClusterName);
+    }
   }
 
   /**
    * Returns the config for all the Helix instances in the cluster.
+   * All the callers are not for segment assignment. No need to use cached accessor.
    */
   public List<InstanceConfig> getAllHelixInstanceConfigs() {
```

Comment:
I think the segment assignment is only calling this method

Reply:
This method is called by `assignInstance` method, which is where we add/update table config.